pkg/operator/clustermemberremovalcontroller/clustermemberremovalcontroller.go

```diff
@@ -0,0 +1,227 @@
+	"github.com/openshift/cluster-etcd-operator/pkg/dnshelpers"
+	"github.com/openshift/cluster-etcd-operator/pkg/etcdcli"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+)
+
+type ClusterMemberRemovalController struct {
+	etcdClient                 etcdcli.EtcdClient
+	masterNodeLister           corev1listers.NodeLister
+	masterMachineLister        machinelistersv1beta1.MachineLister
+	networkLister              configv1listers.NetworkLister
+	hasNetworkInformerSyncedFn func() bool
+	// isMachineAPIFunctionalFn determines if the precondition for this controller is met,
+	// this controller can be run only on a cluster that exposes a functional Machine API
+	isMachineAPIFunctionalFn func() (bool, error)
+}
+
+// NewClusterMemberRemovalController removes an etcd member if the machine and a node for the etcd member is gone and the machine-api is active
+//
+// Note:
+//  since this controller needs to reconcile only master nodes and machine objects
+//  make sure nodeInformer and machineInformer contain only filtered data
+//  otherwise it might be expensive to react to every node update in larger installations
+func NewClusterMemberRemovalController(
+	isMachineAPIFunctional func() (bool, error),
+	etcdClient etcdcli.EtcdClient,
+	masterNodeInformer cache.SharedInformer,
+	masterMachineInformer cache.SharedInformer,
+	masterNodeLister corev1listers.NodeLister,
+	masterMachineLister machinelistersv1beta1.MachineLister,
+	hasNetworkInformerSyncedFn func() bool,
+	networkLister configv1listers.NetworkLister,
+	eventRecorder events.Recorder,
+) factory.Controller {
+	c := &ClusterMemberRemovalController{
+		etcdClient:                 etcdClient,
+		isMachineAPIFunctionalFn:   isMachineAPIFunctional,
+		masterNodeLister:           masterNodeLister,
+		masterMachineLister:        masterMachineLister,
+		networkLister:              networkLister,
+		hasNetworkInformerSyncedFn: hasNetworkInformerSyncedFn,
+	}
+	return factory.New().
+		WithSync(c.sync).
+		ResyncEvery(33*time.Minute). // make it slow since nodes are updated every few minutes
+		WithInformers(
+			masterNodeInformer,
+			masterMachineInformer,
+		).ToController("ClusterMemberRemovalController", eventRecorder.WithComponentSuffix("cluster-member-removal-controller"))
+}
+
+func (c *ClusterMemberRemovalController) sync(ctx context.Context, _ factory.SyncContext) error {
+	if !c.hasNetworkInformerSyncedFn() {
+		return nil
+	}
+	// stop if the machine API is not functional
+	if isFunctional, err := c.isMachineAPIFunctionalFn(); err != nil {
+		return err
+	} else if !isFunctional {
+		return nil
+	}
+	// we are only interested in unhealthy members
+	// because theirs nodes are gone
+	unhealthyMembers, err := c.etcdClient.UnhealthyMembers(ctx)
```

Comment:
I think I am going to replace it with `member list`, I think finding unhealthy members might be expensive, it seems it requires going through the raft.

```
etcdctl endpoint health
https://10.0.139.78:2379 is healthy: successfully committed proposal: took = 8.53226ms
https://10.0.170.28:2379 is healthy: successfully committed proposal: took = 11.234354ms
https://10.0.145.231:2379 is healthy: successfully committed proposal: took = 11.636049ms
```

Reply:
mhm, seem that getting a `member list` might have a similar cost - https://github.com/etcd-io/etcd/blob/96cce208c2cb9e70ac3573b3f76eed7c84d262d1/etcdserver/api/v3rpc/member.go#L91